### PR TITLE
aiohttp 2.x+ client tracing support

### PR DESCRIPTION
**Pull request by thehesiod**
- builds on https://github.com/DataDog/dd-trace-py/pull/259 and supports full circle tracing from client to server and back
- bumps minimum supported aiohttp 2.x version to 2.1, but adds 2.3 to fix unittest
- adds missing aenter/aexit to aiohttp ClientResponse class

### CODE REVIEW

**Comment by thehesiod** [comment]
@palazzem here's the client tracing support I promised. Let me know if you want me re-create this PR to merge to my base aiohttp PR

**Comment by thehesiod** [comment]
btw, reason I added start as a trace is so we can trace redirects.  Perhaps this should be opt-in.  Also this doesn't trace if you read the data by chunks.  Let me know if you'd like to add that.  We don't use it, so perhaps it can be added later.

**Comment by thehesiod** [comment]
@palazzem not sure why this test failed only on py27, ideas?

**Comment by thehesiod** [comment]
I've re-done the implementation since it was broken for concurrent requests, and added support to allow the outer span to last the life of the request.  This also has the benefit of removing usage of private methods.

**Comment by thehesiod** [comment]
ok figured out the service thing, if you don't set a service name the trace will not appear online but you will not get any errors.  Perhaps there there should be a ddtrace core assert if you try tracing w/o a service name?

**Comment by thehesiod** [comment]
btw one of the reasons this was confusing is because I had set Pin.service, but it is not propagated to pin.tracer.trace...perhaps there should be a Pin.trace partial function which binds the service?  I don't know but it's kinda confusing as-is.

**Comment by thehesiod** [comment]
failure looks like a test framework problem

**Comment by palazzem** [comment]
@thehesiod not sure, but this PR seems containing also part of the code of #259 (distributed tracing), right? I think it's better to keep both PR separated and work only with the aiohttp client instrumentation here. Probably it only requires a rebase / cherry-pick. Cool?

**Comment by thehesiod** [comment]
ok I've merged the latest master here.  If this is squash + merged to master it shouldn't matter now right?

**Comment by palazzem** [comment]
@thehesiod don't worry it's fine! at the very end I will squash again this PR so that we have just one commit for your contribution, while keeping the history as a comment. I've found that the CI is red because:
```
    from .middlewares import _SPAN_MIN_ERROR, PARENT_TRACE_HEADER_ID, \
nose.proxy.ImportError: cannot import name '_SPAN_MIN_ERROR'
```
since I've removed `_SPAN_MIN_ERROR` from the implementation because we can't keep that in the next release, while I was more than happy to add the opt-in headers context extract.
Will have a review on that PR and see if we can quickly add it on this week release.

Thank you!

**Comment by asvetlov** [comment]
As discussed with @palazzem during EuroPython we need an aio-libs/aiohttp#1309 implementation.

@palazzem @thehesiod  would you provide a list of required hooks? It could be a good starting point.

**Comment by palazzem** [comment]
Thank you @asvetlov! yes, as we discussed this will help (in general) any kind of instrumentation or logging. Will reach you back very soon with a complete list of hooks we may need to instrument better the `aiohttp` client (and later aio-libs).

Thanks again!

File: ddtrace/contrib/aiohttp/patch.py [hunk 1]
```diff
@@ -1,39 +1,292 @@
...
+
+
+def _set_request_tags(span, url):
+    if (url.scheme == 'http' and url.port == 80) or \
+            (url.scheme == 'https' and url.port == 443):
+        port = ''
+    else:
+        port = url.port
+
+    url_str = '{scheme}://{host}{port}{path}'.format(
```
Comment (asvetlov) [inline ddtrace/contrib/aiohttp/patch.py]: Try `url_str = str(url.with_query(None).with_fragment(None))`
Would you keep userinfo?
If not -- `url_str = str(URL.build(scheme=url.scheme, host=url.host, port=url.port, path=url.path))` could work.
Anyway you've missed a colon between `host` and `port`.

**Comment by thehesiod** [comment]
@asvetlov of course, may take me a bit to get to this, got way behind on things since having our baby :)

**Comment by palazzem** [comment]
@thehesiod following big improvements @asvetlov did in the aiohttp client, I think we can simplify a lot and remove monkey patching here. In that case, we can easily rework the integration + merge that one in our officially supported integrations! We would be very happy to be early adopters!

**Comment by asvetlov** [comment]
My congratulations @thehesiod 

For your information: release date for aiohttp 3.0 is about January.
I love getting any feedback on client tracing API, most likely the feature is not complete yet (was merged in master a couple days ago).

**Comment by asvetlov** [comment]
@pfreixes is the real author of client tracing facility

**Comment by thehesiod** [comment]
@palazzem @asvetlov given people may want to support aiohttp 2.x for awhile should we keep both styles of tracing?

File: ddtrace/contrib/aiohttp/patch.py (hunk 1 above)
Comment (thehesiod) [reply ddtrace/contrib/aiohttp/patch.py]: copy/paste from httplib, so going to keep for find/replace sake

File: ddtrace/contrib/aiohttp/patch.py [hunk 2]
```diff
@@ -1,39 +1,292 @@
...
 
-from ...pin import Pin
 from ddtrace.util import unwrap
 
+from ...propagation.http import HTTPPropagator
+from ...pin import Pin
+from ...ext import http as ext_http
+from ..httplib.patch import should_skip_request
+import aiohttp.client
+from aiohttp.client import URL
```
[the remark this answers is not in the document]
Comment (thehesiod) [reply ddtrace/contrib/aiohttp/patch.py]: updated

File: ddtrace/contrib/aiohttp/patch.py [hunk 3]
```diff
@@ -1,39 +1,292 @@
...
 import wrapt
 
-from ...pin import Pin
 from ddtrace.util import unwrap
 
+from ...propagation.http import HTTPPropagator
+from ...pin import Pin
+from ...ext import http as ext_http
+from ..httplib.patch import should_skip_request
+import aiohttp.client
```
[the remark this answers is not in the document]
Comment (thehesiod) [reply ddtrace/contrib/aiohttp/patch.py]: updated

**Comment by pfreixes** [comment]
Yeps, any plan on using the new aiohttp tracimg system that is available master, fututre 3.0? Just doing the implementation and trying to meet the requirements that you already have should be a good first step, that for sure will provide a relevant feedback.

**Comment by thehesiod** [comment]
Ya I plan on working on that soon. Have a PR in the works to move our latest prod code to ddtrace 10.x. then hopefully switch to aiohttp 3.x with new style tracing. Given ddtrace wants both 2.x and 3.x I'll add a separate impl for 3.x as soon as my PR lands on my side.

**Comment by thehesiod** [comment]
ok, one year later working on 3.x support ;)

**Comment by thehesiod** [comment]
@asvetlov @pfreixes so I'm going over the tracing API and it seems the following is missing:
1. trace for when client is waiting for response headers: https://github.com/aio-libs/aiohttp/blob/v3.3.1/aiohttp/client.py#L387
2. trace for reading from body stream: `StreamReader.read()` yields no traces

Notes
1. Initially I thought a trace was missing for req.send: https://github.com/aio-libs/aiohttp/blob/v3.3.1/aiohttp/client.py#L385 however after tracing through that method and what it calls I don't think it needs to be an async function.  @asvetlov can you confirm and if so remove it as async function?  Will clear up confusion in the future.
2. I think it would be a lot simpler if there was a trace per while loop of redirects, otherwise I have to piece together conn timeout + header send + header response
3. I'm not sure the usefulness of `send_response_chunk_received`
4. seems weird having `send_request_chunk_sent` w/o a received, basically sending request but never got response?  What I see on a request from aiobotocore is: send_request_start -> send_conn_create_start -> send_conn_create_end -> send_request_chunk_sent -> send_request_end

so until the top section is fixed I can't use the new tracing APIs.  Luckily my old aiohttp tracing support still works with a couple minor fixes.

**Comment by asvetlov** [comment]
@thehesiod do you discuss/propose aiohttp changes or datadog client itself?
If the first -- please open an issue in aiohttp bugtracker.

**Comment by thehesiod** [comment]
ok just opened https://github.com/aio-libs/aiohttp/issues/3084, hope it makes sense.  Let me know if I can clarify anything or if I'm not understanding certain things correctly as I missed out on the design phase of this feature.

**Comment by thehesiod** [comment]
just realized that `enable_distributed` needs to be supported on a per-session because otherwise it will mess up tracing with aiobotocore as it would add headers to AWS calls breaking them.

**Comment by thehesiod** [comment]
These failures don't seem to be my fault:
- casandra: `raise NoHostAvailable("Unable to connect to any servers", errors)`
- celery: `pluggy.manager.PluginValidationError: Plugin 'celery' could not be loaded:`

how do I get the tests to pass?

**Comment by jd** [comment]
Don't worry @thehesiod, I'll restart the jobs manually.

File: ddtrace/contrib/aiohttp/patch.py [hunk 4]
```diff
@@ -1,39 +1,374 @@
...
+                span.parent_id = self._self_parent_span_id
+
+            _set_request_tags(span, _get_url_obj(self))
+            result = yield from self.__wrapped__.read(*args, **kwargs)  # noqa: E999
+            span.set_tag(ext_http.STATUS_CODE, self.status)
+            span.set_tag('Length', len(result))
+
+        return result
+
+    if PY_35:
```
[the remark this answers is not in the document]
Comment (thehesiod) [reply ddtrace/contrib/aiohttp/patch.py]: that is >= 3.5: `PY_35 = sys.version_info >= (3, 5)`

File: ddtrace/contrib/aiohttp/middlewares.py [hunk 5]
```diff
@@ -103,14 +121,15 @@ def on_prepare(request, response):
     request_span.finish()
 
 
-def trace_app(app, tracer, service='aiohttp-web'):
+def trace_app(app, tracer, service='aiohttp-web', distributed_tracing=False):
```
[the remark this answers is not in the document]
Comment (thehesiod) [reply ddtrace/contrib/aiohttp/middlewares.py]: can't remember, reverted

**Comment by thehesiod** [comment]
think i need another merge from master to fix this celery issue?

**Comment by jd** [comment]
@irabinovitch at least, this needs some love, the tests still use `nose`, which we removed.

File: ddtrace/contrib/aiohttp/patch.py [hunk 6]
```diff
@@ -1,39 +1,374 @@
...
+            return resp
+
+        @asyncio.coroutine
+        def __aexit__(self, exc_type, exc_val, exc_tb):
+            try:
+                resp = yield from self.__wrapped__.__aexit__(exc_type, exc_val,  # noqa: E999
+                                                             exc_tb)
+            finally:
+                if self._self_have_context and self._self_trace_context:
+                    self._self_span.finish()
```
Comment (jd) [inline ddtrace/contrib/aiohttp/patch.py]: Maybe define a `_finish_span` method that can be used here and in L195 to avoid repeating?

**Comment by thehesiod** [comment]
will get back to this PR asap, thanks for feedback

**Comment by asvetlov** [comment]
Does anybody use aiohttp 2.x now?

**Comment by thehesiod** [comment]
it works in 3.x as well, just it's not using the new aiohttp trace endpoints since it's missing some items: https://github.com/aio-libs/aiohttp/issues/3084

**Comment by asvetlov** [comment]
Thanks for remembering the missing functionality :)

**Comment by mklokocka** [comment]
Any news on this? @thehesiod

**Comment by thehesiod** [comment]
> Any news on this? @thehesiod

unfortunately no, however we're trying to get some more traction on my PRs via a different channel

**Comment by Kyle-Verhoog** [comment]
Thanks so much to @thehesiod for this work! We're going to continue it in #1319 so that we can work collaboratively to get it across the finish line 😄

**Pull request closed without merging**